pkg/offsetguess/offsetguess.go

```diff
@@ -0,0 +1,338 @@
+// +build linux
+
+package offsetguess
+
+import (
+	"fmt"
+	"math/rand"
+	"net"
+	"os"
+	"strconv"
+	"strings"
+	"syscall"
+	"unsafe"
+
+	"github.com/iovisor/gobpf/elf"
+)
+
+type tcpTracerState uint64
+
+const (
+	threshold         = 200
+	thresholdInetSock = 2000
+)
+
+const (
+	uninitialized tcpTracerState = iota
+	checking
+	checked
+	ready
+)
+
+type guessWhat uint64
+
+const (
+	guessSaddr guessWhat = iota
+	guessDaddr
+	guessFamily
+	guessSport
+	guessDport
+	guessNetns
+	guessDaddrIPv6
+)
+
+type tcpTracerStatus struct {
+	status          tcpTracerState
+	pidTgid         uint64
+	what            guessWhat
+	offsetSaddr     uint64
+	offsetDaddr     uint64
+	offsetSport     uint64
+	offsetDport     uint64
+	offsetNetns     uint64
+	offsetIno       uint64
+	offsetFamily    uint64
+	offsetDaddrIPv6 uint64
+	err             byte
+	saddr           uint32
+	daddr           uint32
+	sport           uint16
+	dport           uint16
+	netns           uint32
+	family          uint16
+	daddrIPv6       [4]uint32
+}
+
+type fieldValues struct {
+	saddr     uint32
+	daddr     uint32
+	sport     uint16
+	dport     uint16
+	netns     uint32
+	family    uint16
+	daddrIPv6 [4]uint32
+}
+
+func listen(url, netType string, listenCompleted, closeListen chan struct{}) {
+	l, err := net.Listen(netType, url)
+	if err != nil {
+		panic(err)
+	}
+
+	close(listenCompleted)
+
+	select {
+	case <-closeListen:
+		l.Close()
+		return
+	}
+}
+
+func compareIPv6(a, b [4]uint32) bool {
+	for i := 0; i < 4; i++ {
+		if a[i] != b[i] {
+			return false
+		}
+	}
+	return true
+}
+
+func ownNetNS() (uint64, error) {
+	var s syscall.Stat_t
+	if err := syscall.Stat("/proc/self/ns/net", &s); err != nil {
+		return 0, err
+	}
+	return s.Ino, nil
+}
+
+func ipv6FromUint32Arr(ipv6Addr [4]uint32) net.IP {
+	buf := make([]byte, 16)
+	for i := 0; i < 16; i++ {
+		buf[i] = *(*byte)(unsafe.Pointer((uintptr(unsafe.Pointer(&ipv6Addr[0])) + uintptr(i))))
+	}
+	return net.IP(buf)
+}
+
+func htons(a uint16) uint16 {
+	var x uint32 = 0x01020304
+	switch *(*byte)(unsafe.Pointer(&x)) {
```

Comment:
If you are going to repeat the same code as in Host you may as well leave it as it was.